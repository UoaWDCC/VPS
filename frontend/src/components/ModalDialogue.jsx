import { useEffect, useRef } from "react";
import { createPortal } from "react-dom";

// NOTE: this is unusual on purpose, so we get the full benefits of the native dialog element, notably esc behaviour and auto close for form buttons

function ModalDialog({ title, children, open, onClose, wide = false }) {
  const dialogRef = useRef(null);

  useEffect(() => {
    const dialog = dialogRef.current;
    if (!dialog) return;

    if (open && !dialog.open) dialog.showModal();
    else if (!open && dialog.open) dialog.close();
  }, [open]);

  useEffect(() => {
    const dialog = dialogRef.current;
    if (!dialog) return;

    const handleClose = (e) => {
      onClose?.(e);
    };

    dialog.addEventListener("close", handleClose);
    return () => dialog.removeEventListener("close", handleClose);
  }, [onClose]);

  if (!open) return null;

  return createPortal(
    <dialog ref={dialogRef} id="modal" className="modal font-ibm">
      <form
        method="dialog"
        className={`modal-box flex flex-col text-m max-h-8/10 ${wide ? "max-w-[64rem]" : "overflow-y-visible"}`}
        tabIndex={0}
      >
        <button className="btn btn-sm btn-circle btn-ghost absolute right-2 top-2">
          âœ•
        </button>
        <h3 className="font-bold text-m">{title}</h3>
        <div className="mt-[0.5rem] min-h-0 flex-1 no-scrollbar">
          {children}
        </div>
      </form>
    </dialog>,
    document.getElementById("modal-portal")
  );
}

export default ModalDialog;
