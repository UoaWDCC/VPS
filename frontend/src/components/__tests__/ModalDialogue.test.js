import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import ModalDialogue from "../ModalDialogue";

test("ModalDialogue component snapshot test", () => {
  const component = renderer.create(
    <BrowserRouter>
      <ModalDialogue />
    </BrowserRouter>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
