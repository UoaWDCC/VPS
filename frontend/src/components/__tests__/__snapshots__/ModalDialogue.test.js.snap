// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalDialogue component snapshot test 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div />
  </div>
  <div
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby="customized-dialog-title"
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root Component-root-1"
          id="customized-dialog-title"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6"
          />
        </div>
        <div
          class="MuiDialogContent-root MuiDialogContent-dividers"
        />
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing"
        />
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
